src/typechecker.erl

```diff
@@ -1457,39 +1488,47 @@ type_check_expr(Env, {op, _, '!', Proc, Val}) ->
     {type(any)
     ,union_var_binds(VB1, VB2, Env#env.tenv)
     ,constraints:combine(Cs1, Cs2)};
-type_check_expr(Env, {op, P, 'not', Arg}) ->
+type_check_expr(Env, {op, P, 'not', Arg} = E) ->
     {Ty, VB, Cs1} = type_check_expr(Env, Arg),
     case subtype(Ty, {type, P, boolean, []}, Env#env.tenv) of
         {true, Cs2} ->
             NormTy = normalize(Ty, Env#env.tenv),
             {negate_bool_type(NormTy), VB, constraints:combine(Cs1, Cs2)};
         false ->
-            throw({type_error, non_boolean_argument_to_not, P, Ty})
+            throw({type_error, #{error_type  => type_error,
+                                 expression  => E,
+                                 actual_type => Ty}})
```

Comment:
I like the idea of structuring these types of errors in a uniform way. Here though, some information is lost and there is some assymetry. `E` is the expression `not Arg`. `Ty` is the type of `Arg`. `Arg` is not included in the error and the expected type `boolean()` is not included.

Perhaps we should have `expression => Arg`, `expected_type => type(boolean)` and `context => E`. By context I mean surrounding expression or pattern. Then we have all the information that was encoded in the tag before and can render it as e.g. "The variable Arg in the unary not expression on line 42 has type string() but is expected to have type boolean()".

Apart from this, I think the wrapping tuple is redundant.

Reply:
Rebasing would get increasingly tricky without amending so I unfortunately think it's the way of working that suits me best here.

Reply:
I agree on all your comments.
Wrapping the tuple might be redundant but I know that there is at least one place that catches based on that first element of the tuple so I wasn't sure.